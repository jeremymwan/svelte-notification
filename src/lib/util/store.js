import {writable, get} from 'svelte/store'

function base_store(o = new Set([])) {
	const {set, update, subscribe} = writable(o)

	return {
		getData(store) {
			return globalThis.structuredClone(get(store))
		},
		set(data) {
			set(globalThis.structuredClone(data))
		},
		subscribe,
		update,
	}
}

const KEY = Symbol.for('store.notifications')
const singleton = {}
singleton[KEY] = base_store()

Object.freeze(singleton)

export default singleton[KEY]
