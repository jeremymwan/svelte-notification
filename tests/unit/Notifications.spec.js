import { setTimeout } from 'timers/promises'
import { get } from 'svelte/store'
import { tick } from 'svelte'
import {
	test,
	expect,
	beforeEach,
	vi,
} from 'vitest'
import {
	Notifications,
	acts,
} from '$lib/index.js'

beforeEach(() => {
	document.body.innerHTML = '<main id="xxx"></main>'
})

function doc_query(selector) {
	const node = document.querySelector(selector)
	if (!node) {
		throw new Error(`No element found for selector: ${selector}`)
	}
	return node
}

test('should match snapshot', async () => {
	const target = doc_query('main#xxx')
	new Notifications({ target })
	const notification = { message: 'one', lifetime: 10 }
	acts.add(notification)

	await setTimeout(2000)
	const div = doc_query('div')
	expect(div).toMatchSnapshot()

	const store = acts.getStore()
	console.log('notification', get(store).size)

	const spy = vi.fn()
	const btn = doc_query('button._tadashi_svelte_notification__btn')
	btn.addEventListener('click', () => {
		spy()
		acts.remove(notification)
		console.log('notification', get(store).size)
	})

	const clickEvent = new Event('click')
	btn.dispatchEvent(clickEvent)
	await tick()
	await setTimeout(2000)

	expect(spy, 'clicado').toHaveBeenCalledTimes(1)
	expect(() => doc_query('div')).toThrowError(/No element/)
})

test('size 1', async () => {
	const notification = { message: 'one', lifetime: 10 }
	acts.add(notification)

	const target = doc_query('main#xxx')
	new Notifications({ target })

	await setTimeout(2000)
	const div = doc_query('div')
	expect(div).toMatchSnapshot()
})
