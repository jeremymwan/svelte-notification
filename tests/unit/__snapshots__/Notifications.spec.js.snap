// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<div
  class="_tadashi_svelte_notifications s-DVcJS0o840aA"
  style=""
>
  
  <div
    aria-live="polite"
    class="_tadashi_svelte_notification s-z9u3mgaTh3R0"
    role="status"
  >
    <div
      class="_tadashi_svelte_notification__content s-z9u3mgaTh3R0"
    >
      one
      
    </div>
     
    <button
      aria-label="delete notification"
      class="_tadashi_svelte_notification__btn s-z9u3mgaTh3R0"
    />
  </div>
  <!--&lt;Base&gt;-->
  <!--&lt;Notification&gt;-->
</div>
`;

exports[`size 1 1`] = `
<div
  class="_tadashi_svelte_notifications s-DVcJS0o840aA"
>
  
  <div
    aria-live="polite"
    class="_tadashi_svelte_notification s-z9u3mgaTh3R0"
    role="status"
  >
    <div
      class="_tadashi_svelte_notification__content s-z9u3mgaTh3R0"
    >
      one
      
    </div>
     
    <button
      aria-label="delete notification"
      class="_tadashi_svelte_notification__btn s-z9u3mgaTh3R0"
    />
  </div>
  <!--&lt;Base&gt;-->
  <!--&lt;Notification&gt;-->
</div>
`;
